### PR TITLE
Decouple String bridging functions using ObjC

### PR DESCRIPTION
**Pull request by Catfish-Man**

### CODE REVIEW

**Comment by Catfish-Man** [comment]
@swift-ci please benchmark

**Comment by Catfish-Man** [comment]
oops looks like correctness tests are failing locally

**Comment by Catfish-Man** [comment]
That's pretty good actually! A few speedups, and the regressions are mostly manageable looking.

**Comment by Catfish-Man** [comment]
@swift-ci please benchmark

**Comment by Catfish-Man** [comment]
Not all tests are passing, but it's close enough that we can remeasure to see if the new approach is promising

**Comment by Catfish-Man** [comment]
Specifically there's failures in these two: 
    Swift(macosx-x86_64) :: stdlib/StringAPI.swift
    Swift(macosx-x86_64) :: stdlib/TestJSONEncoder.swift

**Comment by Catfish-Man** [comment]
I think the benchmarks hung :( gonna try 'em locally

**Comment by Catfish-Man** [comment]
@swift-ci please benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please benchmark

**Comment by Catfish-Man** [comment]
Oh ho ho that is promising. I think this approach is gonna work. Still need to figure out what I did to ObjectiveCBridgeASCIIStringFromFile and NSStringConversion.*UTF8 though. Although once the correctness issues are fixed and I clean up the code this is probably almost good enough to land.

**Comment by Catfish-Man** [comment]
Gonna run one more to see if some of that is noise, a few of those shouldn't have changed

**Comment by Catfish-Man** [comment]
@swift-ci please benchmark

**Comment by Catfish-Man** [comment]
Pretty consistent then… fascinating. I wonder how I did that.

**Comment by Catfish-Man** [comment]
Hm. I can't reproduce the ObjectiveCBridgeASCIIStringFromFile regression locally.

**Comment by Catfish-Man** [comment]
Ok the NSStringConversion.LongUTF8 regression is because that does O(n) characterAtIndex calls

**Comment by Catfish-Man** [comment]
@swift-ci please benchmark

**Comment by Catfish-Man** [comment]
Ugh I think I broke more stuff heh. Gonna spend some time fixing correctness now.

**Comment by Catfish-Man** [comment]
@swift-ci please benchmark

**Comment by Catfish-Man** [comment]
Well, ObjectiveCBridgeASCIIStringFromFile and NSStringConversion.LongUTF8 are fixed, but I regressed everything else. Sigh…

**Comment by Catfish-Man** [comment]
@swift-ci please benchmark

**Comment by Catfish-Man** [comment]
forgot to push >.<

**Comment by Catfish-Man** [comment]
@swift-ci please benchmark

**Comment by Catfish-Man** [comment]
Definitely definitely screwed up tagged pointer handling. I think I improved the rest, but it's hard to tell since that throws things off.

**Comment by Catfish-Man** [comment]
Figured it out. That SPI takes CF encodings instead of NS encodings.

**Comment by Catfish-Man** [comment]
@swift-ci please test

**Comment by Catfish-Man** [comment]
@swift-ci please benchmark

**Comment by Catfish-Man** [comment]
Ok that’s a lot closer. Progress!

**Comment by Catfish-Man** [comment]
@swift-ci please test

**Comment by Catfish-Man** [comment]
@swift-ci please benchmark

**Comment by Catfish-Man** [comment]
This run will be slower for tagged strings again. I have a plan for fixing that in a subsequent commit, but the approach I used to get the 1.33x win above fails in edge cases (specifically: embedded \0s in NSStrings)

**Comment by Catfish-Man** [comment]
!!

It's still faster :D

**Comment by Catfish-Man** [comment]
@swift-ci please test

File: stdlib/public/SwiftShims/CoreFoundationShims.h [hunk 1]
```diff
@@ -55,44 +55,6 @@ typedef unsigned char _swift_shims_Boolean;
 typedef __swift_uint16_t _swift_shims_UniChar;
 typedef __swift_uint8_t _swift_shims_UInt8;
 
-// Buffer is nullable in case the string is zero-length.
-SWIFT_RUNTIME_STDLIB_API
```
Comment (Catfish-Man) [inline stdlib/public/SwiftShims/CoreFoundationShims.h]: If something is somehow depending on these directly, we can reimplement them using dlsym, or the same tricks but in a .mm file, but I think we should see if we can get away with removing them.

**Comment by Catfish-Man** [comment]
Filed https://bugs.swift.org/browse/SR-11249 to cover the janky @_effects workaround

**Comment by Catfish-Man** [comment]
@swift-ci please test

**Comment by Catfish-Man** [comment]
The possible minor regression on NSStringConversion.Medium, NSStringConversion.Long, and ObjectiveCBridgeASCIIStringFromFile is at least partly because -[__NSCFString _fastCStringContents:] isn't inlining the CF bits it calls into, which is something we could improve on the ObjC side later. Inlining getSharedUTF8Start() into sharedUTF8 on our end would buy about 4% and should be safe, but isn't a big deal.

**Comment by Catfish-Man** [comment]
Huh… weird. Those tests weren't failing locally. Oh well, something to investigate tomorrow.

**Comment by Catfish-Man** [comment]
@swift-ci please test

File: stdlib/public/core/StringBridge.swift [hunk 2]
```diff
@@ -17,34 +17,113 @@ import SwiftShims
 internal typealias _CocoaString = AnyObject
 
 #if _runtime(_ObjC)
+
 // Swift's String bridges NSString via this protocol and these
 // variables, allowing the core stdlib to remain decoupled from
 // Foundation.
 
+@objc internal protocol _StringSelectorHolder : _NSCopying {
```
Comment (nvzqz) [inline stdlib/public/core/StringBridge.swift]: If this isn't being used externally, why is it `internal`?

File: stdlib/public/core/StringBridge.swift [hunk 3]
```diff
@@ -17,34 +17,113 @@ import SwiftShims
...
+ compiler doesn't know about, via the protocol above. In order to get good
+ performance, you need a double indirection like this:
+ 
+  func a -> _objc -> func a'
+ 
+ because any refcounting @_effects on 'a' will be lost when _objc breaks ARC's
+ knowledge that the _CocoaString and _StringSelectorHolder are the same object.
+ */
+@inline(__always)
+internal func _objc(_ str: _CocoaString) -> _StringSelectorHolder {
```
Comment (nvzqz) [inline stdlib/public/core/StringBridge.swift]: Does declaring this as a freestanding function benefit us in any way? At least for organization, it'd be nice to have this as an extension on `_CocoaString`.

File: stdlib/public/core/StringBridge.swift [hunk 4]
```diff
@@ -17,34 +17,113 @@ import SwiftShims
...
+  return unsafeBitCast(str, to: _StringSelectorHolder.self)
+}
+
+@_effects(releasenone)
+private func _copyNSString(_ str: _StringSelectorHolder) -> _CocoaString {
```
Comment (nvzqz) [inline stdlib/public/core/StringBridge.swift]: This only gets called once so why not just inline the call to `copy(with:)`?

File: stdlib/public/core/StringBridge.swift [hunk 5]
```diff
@@ -55,68 +134,100 @@ internal func _cocoaStringCopyCharacters(
...
 ) {
-  _swift_stdlib_CFStringGetCharacters(
-    source,
-    _swift_shims_CFRange(location: range.lowerBound, length: range.count),
-    destination)
+  _NSStringGetCharacters(from: _objc(source), range: range, into: destination)
+}
+
+@_effects(readonly)
+private func _NSStringGetCharacter(
```
Comment (nvzqz) [inline stdlib/public/core/StringBridge.swift]: Similarly with suggestion about `_copyNSString`, perhaps this body should get inlined into `_cocoaStringSubscript`.

File: stdlib/public/core/StringBridge.swift (hunk 4 above)
Comment (Catfish-Man) [reply stdlib/public/core/StringBridge.swift]: (we discussed this and the one below in person)

File: stdlib/public/core/StringBridge.swift (hunk 3 above)
Comment (Catfish-Man) [reply stdlib/public/core/StringBridge.swift]: also discussed in person

File: stdlib/public/core/StringBridge.swift (hunk 2 above)
Comment (Catfish-Man) [reply stdlib/public/core/StringBridge.swift]: Fixed

**Review by lorentey — APPROVED** [review]
LGTM!

File: stdlib/public/core/StringUTF8View.swift [hunk 6]
```diff
@@ -505,7 +506,7 @@ extension String.UTF8View {
       // _cocoaStringUTF8Count gave us the scalar aligned count, but we still
       // need to compensate for sub-scalar indexing, e.g. if `i` is in the
       // middle of a two-byte UTF8 scalar.
-      let refinedCount = count - (i.transcodedOffset + j.transcodedOffset)
+      let refinedCount = (count - i.transcodedOffset) + j.transcodedOffset
```
[the remark this answers is not in the document]
Comment (Catfish-Man) [reply stdlib/public/core/StringUTF8View.swift]: Yeah my next task is making more tests for this, since it exposed a hole in our existing coverage.

**Comment by Catfish-Man** [comment]
@swift-ci please test and merge

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test and merge

File: stdlib/public/core/StringUTF8View.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (stephentyrone) [reply stdlib/public/core/StringUTF8View.swift]: File under "bugs that a type system should actually save us from, but we still found a way to achieve"

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test and merge

**Comment by shahmishal** [comment]
@swift-ci please smoke test and merge

**Comment by Catfish-Man** [comment]
oddly it's not noticing my force push to fix that build failure, but let's try this again

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test and merge

**Comment by Catfish-Man** [comment]
Um. Ok… apparently my commits really aren't showing up at the GitHub end? but are in git? how is that possible

**Comment by Catfish-Man** [comment]
…now it worked. Bizarre

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test and merge

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test and merge

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test and merge

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test and merge

**Pull request merged**